commit 62e8b17ffc2ff0b0e29d5e05a18570c3e70b35ff
Author: Lukas Wunner <lukas@wunner.de>
Date:   Sat Mar 11 15:40:07 2023 +0100

    PCI/DOE: Provide synchronous API and use it internally
    
    The DOE API only allows asynchronous exchanges and forces callers to
    provide a completion callback.  Yet all existing callers only perform
    synchronous exchanges.  Upcoming commits for CMA (Component Measurement
    and Authentication, PCIe r6.0 sec 6.31) likewise require only
    synchronous DOE exchanges.
    
    Provide a synchronous pci_doe() API call which builds on the internal
    asynchronous machinery.
    
    Convert the internal pci_doe_discovery() to the new call.
    
    The new API allows submission of const-declared requests, necessitating
    the addition of a const qualifier in struct pci_doe_task.
    
    Tested-by: Ira Weiny <ira.weiny@intel.com>
    Signed-off-by: Lukas Wunner <lukas@wunner.de>
    Reviewed-by: Ming Li <ming4.li@intel.com>
    Reviewed-by: Ira Weiny <ira.weiny@intel.com>
    Reviewed-by: Davidlohr Bueso <dave@stgolabs.net>
    Reviewed-by: Dan Williams <dan.j.williams@intel.com>
    Reviewed-by: Jonathan Cameron <Jonathan.Cameron@huawei.com>
    Acked-by: Bjorn Helgaas <bhelgaas@google.com>
    Link: https://lore.kernel.org/r/0f444206da9615c56301fbaff459c0f45d27f122.1678543498.git.lukas@wunner.de
    Signed-off-by: Dan Williams <dan.j.williams@intel.com>

diff --git a/include/linux/pci-doe.h b/include/linux/pci-doe.h
index 43765eaf2342..5dcd54f892e5 100644
--- a/include/linux/pci-doe.h
+++ b/include/linux/pci-doe.h
@@ -49,7 +49,7 @@ struct pci_doe_mb;
  */
 struct pci_doe_task {
 	struct pci_doe_protocol prot;
-	__le32 *request_pl;
+	const __le32 *request_pl;
 	size_t request_pl_sz;
 	__le32 *response_pl;
 	size_t response_pl_sz;
@@ -78,4 +78,8 @@ struct pci_doe_mb *pcim_doe_create_mb(struct pci_dev *pdev, u16 cap_offset);
 bool pci_doe_supports_prot(struct pci_doe_mb *doe_mb, u16 vid, u8 type);
 int pci_doe_submit_task(struct pci_doe_mb *doe_mb, struct pci_doe_task *task);
 
+int pci_doe(struct pci_doe_mb *doe_mb, u16 vendor, u8 type,
+	    const void *request, size_t request_sz,
+	    void *response, size_t response_sz);
+
 #endif